File: solution1.py
import os
import time
import multiprocessing


def find_new_line(chunks: list[list[int]], file_with_ip: str) -> list[list[int]]:
    # go through each pair of [start, end]
    # adjust end to the closest start of the line
    # shift current end and next start

    with open(file_with_ip, 'br') as file:
        for i in range(len(chunks) - 1):
            shift = 0
            file.seek(chunks[i][1])
            while file.read(1) != b'\n':
                shift += 1
            chunks[i][1] += shift + 1
            chunks[i + 1][0] += shift + 1
    return chunks


def split_file_to_chunks(file_with_ip: str, process_count: int) -> list[list[int]]:
    # check filesize and create several ranges
    # to read file in chunks

    filesize = os.path.getsize(file_with_ip)
    chunks_len = filesize // process_count
    chunks = [[i, i + chunks_len] for i in range(0, filesize - process_count, chunks_len)]
    chunks[-1][1] = filesize
    chunks = find_new_line(chunks, file_with_ip)
    return chunks


def fill_bitmap(file_with_ip: str, chunk: list[int, int]) -> None:
    global bitmap
    start_pos = chunk[0]
    end_pos = chunk[1]
    with open(file_with_ip, 'r') as file:
        # move pointer to starting position
        file.seek(start_pos)
        while start_pos < end_pos:
            # read line by line in current chunk
            curr_line = file.readline()
            start_pos += len(curr_line)
            # convert ip to integer
            octets = curr_line.strip().split(".")
            ip_to_number = (int(octets[0]) << 24) | (int(octets[1]) << 16) | (int(octets[2]) << 8) | (int(octets[3]))
            # set corresponding bit in the bitmap
            bitmap[ip_to_number // 32] |= (1 << (ip_to_number % 32))


def count_unique_ips(start_pos: int, end_pos: int) -> int:
    # go through the bitmap
    # and count set bits
    global bitmap
    return sum(1 for x in range(start_pos, end_pos) if (bitmap[x // 32] & (1 << (x % 32))))


def main() -> int:
    start_time = time.perf_counter()
    # little config
    # file_with_ip = 'test_data3.txt' # 1000
    # file_with_ip = "test_data.txt" # 9988184
    file_with_ip = "test_data2.txt"  # 98845647
    process_count = multiprocessing.cpu_count() // 2
    names = [file_with_ip] * process_count

    # split file into chunks [a,b] ]to read in separate processes
    # then modify the ranges, so that a is always at the beginning of a new line
    chunks_ranges = split_file_to_chunks(file_with_ip, process_count)

    # read file and fill bitmap to count unique ip
    with multiprocessing.Pool(processes=process_count) as pool:
        pool.starmap(fill_bitmap, zip(names, chunks_ranges))
    end_time = time.perf_counter()

    print("done reading file and filling bitmap, took: ", end_time - start_time)

    start_time = time.perf_counter()

    # create chunks of [a,b] to split entire bitmap
    chunks_len = 2 ** 32 // process_count
    chunks_ranges = [[i, i + chunks_len] for i in range(0, 2 ** 32 - process_count, chunks_len)]
    chunks_ranges[-1][1] = 2 ** 32

    with multiprocessing.Pool(processes=process_count) as pool:
        chunk_sum = pool.starmap(count_unique_ips, chunks_ranges)
    res = sum(chunk_sum)

    end_time = time.perf_counter()
    print("done counting unique number, took: ", end_time - start_time)
    return res


bitmap = multiprocessing.RawArray('i', (2 ** 27))

if __name__ == "__main__":
    result = main()
    print(result)

File: solution1.go
package main

import (
	"bufio"
	"fmt"
	"io"
	"math/bits"
	"os"
	"strconv"
	"strings"
	"sync"
	"time"
)

// go through each pair of [start, end]
// adjust end to the closest start of the line
// shift current end and next start
func findNewLine(chunks [][]int64, fileWithIP string) {
	file, err := os.Open(fileWithIP)
	if err != nil {
		panic(err)
	}
	defer file.Close()
	reader := bufio.NewReader(file)

	for i := 0; i < len(chunks)-1; i++ {
		file.Seek(int64(chunks[i][1]), io.SeekStart)
		line, _ := reader.ReadString('\n')
		reader.Discard(reader.Buffered())
		shift := int64(len(line))
		chunks[i][1] += shift
		chunks[i+1][0] += shift
	}
}

// create pairs of chunks depending on the filesize
func splitFileToChunks(fileWithIP string, processCount int) [][]int64 {
	fileInfo, err := os.Stat(fileWithIP)
	if err != nil {
		panic(err)
	}
	filesize := fileInfo.Size()

	chunksLen := filesize / int64(processCount)
	chunks := make([][]int64, processCount)
	for i := 0; i < processCount-1; i++ {
		chunks[i] = []int64{int64(i) * chunksLen, int64(i+1) * chunksLen}
	}
	chunks[processCount-1] = []int64{int64(processCount-1) * chunksLen, filesize}
	findNewLine(chunks, fileWithIP)
	return chunks
}

// read each IP
// convert it into int
// set the corrsponding bit to 1
func fillBitmap(fileWithIPName string, chunk []int64, bitmap []byte) { //  wg *sync.WaitGroup
	var start int64 = chunk[0]
	var end int64 = chunk[1]
	ipFile, _ := os.Open(fileWithIPName)
	defer ipFile.Close()

	ipFile.Seek(int64(chunk[0]), io.SeekStart)
	reader := bufio.NewReader(ipFile)
	for start < end {
		ip, err := reader.ReadString('\n')
		if err == io.EOF {
			break
		}
		start += int64(len(ip))
		ip = strings.TrimSpace(ip)
		parts := strings.Split(string(ip), ".")
		var octets [4]int
		for i := 0; i < 4; i++ {
			octets[i], err = strconv.Atoi(parts[i])
		}
		convertedIP := uint32(octets[0]<<24 | octets[1]<<16 | octets[2]<<8 | octets[3])
		bitmap[convertedIP/8] |= (1 << (convertedIP % 8))

	}
}

func countUniqueIPs(chunk []int64, bitmap []byte) uint32 {
	startPos := chunk[0] / 8
	endPos := chunk[1] / 8
	var sum uint32
	for i := startPos; i < endPos; i++ {
		sum += uint32(bits.OnesCount8(bitmap[i]))
	}

	return sum
}

// create pairs to go through each of 2^32 nums
// to check later for set bits
func splitRangeToChunks(processCount int) [][]int64 {
	chunks := make([][]int64, processCount)
	maxLen := int64((1 << 32))
	chunksLen := (maxLen) / int64(processCount)
	for i := 0; i < processCount-1; i++ {
		chunks[i] = []int64{int64(i) * chunksLen, int64(i+1) * chunksLen}
	}
	chunks[processCount-1] = []int64{int64(processCount-1) * chunksLen, maxLen + 1}

	return chunks
}

func main() {
	// config
	// fileWithIP := "test_data.txt"
	fileWithIP := "ip_addresses"
	processCount := 10
	var bitmap []byte = make([]byte, 1<<29)

	chunks := splitFileToChunks(fileWithIP, processCount)
	var wg sync.WaitGroup
	start := time.Now()
	wg.Add(processCount)
	for i := 0; i < processCount; i++ {
		go func(i int) {
			defer wg.Done()
			fillBitmap(fileWithIP, chunks[i], bitmap)
		}(i)
	}

	wg.Wait()
	end := time.Now()
	fmt.Println("done reading file and filling bitmap, took: ", end.Sub(start))

	rangeChunks := splitRangeToChunks(processCount)
	var result uint32
	wg.Add(processCount)
	for i := 0; i < processCount; i++ {
		go func(i int) {
			defer wg.Done()
			result += countUniqueIPs(rangeChunks[i], bitmap)

		}(i)
	}
	wg.Wait()
	fmt.Println(result)
}
